### ISSUE TITLE
error="http: proxy error: context canceled" / "Data proxy error" for Loki Data Source

### ISSUE THREAD

**Issue by borisyukd** [grafana/grafana#31638]
**What happened**:
I have Grafana Loki as Data Source. It could be connected directly (Grafana -> Loki) or through Nginx proxy (Grafana -> Nginx -> Loki). The issue exists in both cases.
Using Expore Grafana interface I select more than 1h time range and request fails with 60 second timeout and HTTP 504 error code.

**What you expected to happen**:

Grafana should return valid data from Loki with HTTP 200 code

**How to reproduce it (as minimally and precisely as possible)**:

- install Grafana
- install Loki
- setup Loki as Grafana Data Source
- setup Grafana '[dataproxy]' timeouts
- load enough data to Loki to get request execution with more than 60 seconds
- try to make a request via Grafana Explore interface 

**Anything else we need to know?**:

Grafana dataproxy configuration from grafana.ini
```
[dataproxy]
timeout = 240
keep_alive_seconds = 240
idle_conn_timeout_seconds = 300
```
Grafana log
```
t=2021-03-03T12:46:16+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname="%USERNAME%" path=/api/datasources/proxy/1/loki/api/v1/query_range remote_addr=%SERVER_IP% referer="https://%GRAFANA_DOMAIN%/explore?orgId=1&left=%5B%22now-3h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bsource%3D%5C%22stdout%5C%22%7D%22%7D%5D" error="http: proxy error: context canceled"
t=2021-03-03T12:46:16+0000 lvl=eror msg="Request Completed" logger=context userId=2 orgId=1 uname="%USERNAME%" method=GET path=/api/datasources/proxy/1/loki/api/v1/query_range status=502 remote_addr=%SERVER_IP% time_ms=59993 size=0 referer="https://%GRAFANA_DOMAIN%/explore?orgId=1&left=%5B%22now-3h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bsource%3D%5C%22stdout%5C%22%7D%22%7D%5D"
```

Grafana UI nginx.log
```
%CLIENT_IP%   443     %SERVER_IP%    2021-03-03T12:46:16+00:00       forward-to 127.0.0.1:3000       GET /api/datasources/proxy/1/loki/api/v1/query_range?direction=BACKWARD&limit=1000&query=%7Bsource%3D%22stdout%22%7D%20%7C%3D%20%22some%20text%22&start=1614764714000000000&end=1614775515000000000&step=5 HTTP/1.1     499     0       59.993  -       https://%GRAFANA_DOMAIN%/explore?orgId=1&left=%5B%22now-3h%22,%22now%22,%22Loki%22,%7B%22expr%22:%22%7Bsource%3D%5C%22stdout%5C%22%7D%22%7D%5D       Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36       %COOKIE%    %GRAFANA_DOMAIN%    TLSv1.2 ECDHE-RSA-AES256-GCM-SHA384
```

Loki API Nginx log
```
127.0.0.1       443     127.0.0.1       2021-03-03T12:46:16+00:00       forward-to 127.0.0.1:3100       GET /loki/api/v1/query_range?direction=BACKWARD&limit=1000&query=%7Bsource%3D%22stdout%22%7D%20%7C%
3D%20%22some%20text%22&start=1614764714000000000&end=1614775515000000000&step=5 HTTP/1.1     499     0       59.993  -       -       Grafana/7.4.3   -       %LOKI_DOMAIN%       TLSv1.2 ECDHE-RSA-AES2
56-GCM-SHA384
```

Loki log
```
Mar  3 12:46:17 %HOSTNAME% loki-linux-arm64: level=info ts=2021-03-03T12:46:17.836548771Z caller=metrics.go:83 org_id=liqpay traceID=3026a26947761ad latency=slow query="{source=\"stdout\"} |= \"som
e text\"" query_type=filter range_type=range length=3h0m1s step=5s duration=1m1.181312131s status=200 throughput=158MB total_bytes=9.7GB
```

Grafana UI response
```
<html> <head><title>504 Gateway Time-out</title></head> <body> <center><h1>504 Gateway Time-out</h1></center> </body> </html> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page -->
```
![image](https://user-images.githubusercontent.com/5219872/109810185-e3155580-7c31-11eb-8226-cbd7e164470c.png)

Loki API response
```
curl -ks 'https://user:password@%LOKI_DOMAIN%/loki/api/v1/query_range?direction=BACKWARD&limit=1000&query=%7Bsource%3D%22stdout%22%7D%20%7C%3D%20%22some%20text%22&start=1614764714000000000&end=1614775515000000000&step=5' | jq
{
  "status": "success",
  "data": {
    "resultType": "streams",
    "result": [],
    "stats": {
      "summary": {
        "bytesProcessedPerSecond": 158368656,
        "linesProcessedPerSecond": 211030,
        "totalBytesProcessed": 10899662783,
        "totalLinesProcessed": 14524077,
        "execTime": 68.824621339
      },
      "store": {
        "totalChunksRef": 1433,
        "totalChunksDownloaded": 1433,
        "chunksDownloadTime": 0.640938809,
        "headChunkBytes": 0,
        "headChunkLines": 0,
        "decompressedBytes": 6640665712,
        "decompressedLines": 8835950,
        "compressedBytes": 1225490760,
        "totalDuplicates": 0
      },
      "ingester": {
        "totalReached": 1,
        "totalChunksMatched": 18,
        "totalBatches": 0,
        "totalLinesSent": 0,
        "headChunkBytes": 0,
        "headChunkLines": 0,
        "decompressedBytes": 4258997071,
        "decompressedLines": 5688127,
        "compressedBytes": 771968768,
        "totalDuplicates": 0
      }
    }
  }
}
```

**Environment**:
- Grafana version: ```Version 7.4.3 (commit: 010f20c1c8, branch: HEAD)```
- Data source type & version:
```
# loki  -version
2021-03-03 12:56:52.151078 I | proto: duplicate proto type registered: ingester.Series
loki, version 2.1.0 (branch: HEAD, revision: 1b79df37)
  build user:       root@b8941a81fe55
  build date:       2020-12-24T14:08:17Z
  go version:       go1.15.3
  platform:         linux/arm64
```
- OS Grafana is installed on:
```
Amazon Linux release 2 (Karoo)
Linux %HOSTNAME% 4.14.214-160.339.amzn2.aarch64 #1 SMP Sun Jan 10 05:53:19 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```
- User OS & Browser:
```
Google Chrome Version 88.0.4324.150 (Official Build) (64-bit)

Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```
- Grafana plugins: No additional plugins installed
- Others: Please find some details here: https://github.com/grafana/grafana/issues/24611

**Comment by borisyukd** [comment]
The issue exists in 7.5-beta https://github.com/grafana/grafana/commit/a1e227638c0f243c81de320db7df9fcac5b7ad00

**How to reproduce it**
- Clone project
- Apply patch if needed
```
# git diff
diff --git a/Dockerfile b/Dockerfile
index f373623dd4..3408f19e52 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -4,6 +4,7 @@ WORKDIR /usr/src/app/
 
 COPY package.json yarn.lock ./
 COPY packages packages
+RUN apk add --no-cache ca-certificates git
 
 RUN yarn install --pure-lockfile --no-progress
 
@@ -73,7 +74,7 @@ RUN export GF_GID_NAME=$(getent group $GF_GID | cut -d':' -f1) && \
     chown -R "grafana:$GF_GID_NAME" "$GF_PATHS_DATA" "$GF_PATHS_HOME/.aws" "$GF_PATHS_LOGS" "$GF_PATHS_PLUGINS" "$GF_PATHS_PROVISIONING" && \
     chmod -R 777 "$GF_PATHS_DATA" "$GF_PATHS_HOME/.aws" "$GF_PATHS_LOGS" "$GF_PATHS_PLUGINS" "$GF_PATHS_PROVISIONING"
 
-COPY --from=go-builder /go/src/github.com/grafana/grafana/bin/linux-amd64/grafana-server /go/src/github.com/grafana/grafana/bin/linux-amd64/grafana-cli ./bin/
+COPY --from=go-builder /go/src/github.com/grafana/grafana/bin/linux-arm64/grafana-server /go/src/github.com/grafana/grafana/bin/linux-arm64/grafana-cli ./bin/
 COPY --from=js-builder /usr/src/app/public ./public
 COPY --from=js-builder /usr/src/app/tools ./tools
```
- Build grafana-server image ```docker build -t grafana-build .```
- Run Grafana ```docker run --rm -p3000:3000 grafana-build:latest```
- Authorize to Grafana UI
- Add Loki Data Source with enough data
- Create query like ```{label="name"} |= "any text"```
- Get Grafana 504 Gateway timeout  response

**Comment by robbertvdg** [comment]
I have a similar issue. I see the same errors, however it happens after 50 seconds instead of 60.
```
2021-04-09 16:39:39 | t=2021-04-09T14:39:39+0000 lvl=eror msg="Request Completed" logger=context userId=2 orgId=1 uname=xxx method=GET path=/api/datasources/proxy/1/loki/api/v1/query_range status=502 remote_addr=xxx time_ms=50000 size=0 referer=xxx
2021-04-09 16:39:39 | t=2021-04-09T14:39:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=xxx path=/api/datasources/proxy/1/loki/api/v1/query_range remote_addr=xxx referer=xxx error="http: proxy error: context canceled"
```
I was able to query the exact same query from the grafana container to the loki gateway using wget, resolving after +/- 70 seconds. 
I modified dataproxy timeouts to 10 minutes.

**Comment by ivanahuckova** [comment]
You might be hitting **query_timeout**. You can [change it in your config](https://grafana.com/docs/loki/latest/configuration/#querier_config). Also it doesn't seem to be related to Grafana and I have found the same issue in Loki repo https://github.com/grafana/loki/issues/3244. So closing this, but feel free to reopen if you think that problem is with Grafana.

**Comment by robbertvdg** [comment]
It has to be a grafana config, since I was able to query the exact same query from the grafana container (NOT the UI) to the loki gateway using wget, resolving after +/- 70 seconds.

**Comment by borisyukd** [comment]
The same for me. query_timeout was set to 3m. You can find in initial details that Loki API returns data correctly but not the Grafana UI.
Here is Loki `querier` block configuration:
```
querier:
  query_timeout: 3m
  engine:
    timeout: 5m
```

**Comment by ivanahuckova** [comment]
What about your Grafana `timeout` setting in custom.ini file? 
```
# How long the data proxy waits before timing out, default is 30 seconds.
# This setting also applies to core backend HTTP data sources where query requests use an HTTP client with timeout set.
;timeout = 30
```

Btw, I have [found this PR](https://github.com/grafana/grafana/pull/31871) that will be part of the next release - we have added UI to add custom timeouts for each data source?

**Comment by borisyukd** [comment]
@ivanahuckova thanks for the link.
I tested [new version](https://github.com/grafana/grafana/commit/d66a5e65a42e675b33e56cc7930402283e677216) and I confirm that Grafana UI does not fail to render long request if Loki was configured with correct datasource timeout (for instance, 300s). 

### Loki API info:
```
Apr 23 14:11:09 %HOSTNAME% loki-linux-arm64[21960]: level=info ts=2021-04-23T14:11:09.128576316Z caller=metrics.go:83 org_id=%ORG_ID% traceID=73141a64d4740ee5 latency=slow query="{service=\"a service\"} |~ \"123123123\"" query_type=filter range_type=range length=24h0m1s step=2m0s duration=2m31.041618349s status=200 throughput=131MB total_bytes=20GB
```
### Grafana Inspector info:
![Screenshot (90)](https://user-images.githubusercontent.com/5219872/115884402-8aa34d00-a457-11eb-9828-64aa539fc7fe.png)

**Comment by robbertvdg** [comment]
That is good to hear! Is this the latest release? And did you change the datasource timeout with the new option?

**Comment by borisyukd** [comment]
It`s the latest commit from **master** branch that contains PR that @ivanahuckova mentioned in https://github.com/grafana/grafana/issues/31638#issuecomment-825607645.
You may compile it from source or wait for the next release on my opinion.
When you start Grafana you will see new **timeout** parameter in Loki datasource settings.
![Screenshot (91)](https://user-images.githubusercontent.com/5219872/115895907-9563df00-a463-11eb-9dff-0a7cf57eebad.png)

**Comment by alex-berger** [comment]
Same here with Grafana v7.5.1. Executing long-running Loki queries from within Grafana (Explorer) runs into timeout after approximately 16 seconds, while the same query executed via `logcli` works fine and does not timeout.

**Comment by sergkolomiiets** [comment]
@borisyukd
Still doesn't work for me.
I created docker image from main branch, from 7.5.6 tag and got after 60 seconds:

**error="http: proxy error: context canceled"**

dataproxy.timeout = 300
Datasource timeout (loki) = 300

Could you please provide commit hash with working timeout parameter so I can check also?

@ivanahuckova

**Comment by ivanahuckova** [comment]
@sergkolomiiets Do you have Grafana behind the proxy? Maybe your connection between the browser and the grafana backend was interrupted while waiting for the proxied request to return.

**Comment by robbertvdg** [comment]
Is the feature released in 7.5.6?

**Comment by ivanahuckova** [comment]
@robbertvdg No it wasn't. It is going to be part of the next Grafana 8 release.

**Comment by sergkolomiiets** [comment]
@ivanahuckova 
You were right. Load balancer idle_timeout.
Thanks!

**Comment by robbertvdg** [comment]
I think loadbalancer might be the issue here as well, good call.

**Comment by yasharne** [comment]
@ivanahuckova I have this problem on `8.1.4`, timeout is set to 300s but I still get 504 after 30seconds

**Comment by hgranillo** [comment]
I had a similar issue to the one described here and I got here by googling the data proxy error, this might help someone in a similar situation.

I have the feature toggle `fullRangeLogsVolume` enabled and when using the Explore (with Loki datasource) I got consistent timeouts trying to load the Log volume metric graph.

![image](https://user-images.githubusercontent.com/16230136/147564386-ebe7e584-f989-4e21-8c83-53bb42a8b79a.png)

Setting the dataproxy timeout to something higher (120) than the default value (10) did the trick.
